### PR TITLE
bpo-40232: Update functions to use _PyThread_at_fork_reinit

### PR DESCRIPTION
**Pull request by corona10**
https://bugs.python.org/issue40232

### CODE REVIEW

**Comment by corona10** [comment]
@vstinner Please take a look :)

File: Python/ceval.c [hunk 1]
```diff
@@ -427,10 +427,16 @@ _PyEval_ReInitThreads(_PyRuntimeState *runtime)
     take_gil(tstate);
 
     struct _pending_calls *pending = &tstate->interp->ceval.pending;
+#ifdef HAVE_FORK
+    if (_PyThread_at_fork_reinit(&pending->lock) < 0) {
+        Py_FatalError("Can't initialize threads for pending calls");
+    }
+#else
     pending->lock = PyThread_allocate_lock();
```
Comment (vstinner) [inline Python/ceval.c]: With _PyThread_at_fork_reinit() call, ``pending->lock = PyThread_allocate_lock()`` is no longer needed!

File: Python/pystate.c [hunk 2]
```diff
@@ -147,6 +147,24 @@ _PyRuntimeState_ReInitThreads(_PyRuntimeState *runtime)
     PyMemAllocatorEx old_alloc;
     _PyMem_SetDefaultAllocator(PYMEM_DOMAIN_RAW, &old_alloc);
 
+#ifdef HAVE_FORK
```
Comment (vstinner) [inline Python/pystate.c]: _PyRuntimeState_ReInitThreads() is only called by PyOS_AfterFork_Child(): so only if HAVE_FORK is defined.

I suggest to only defined _PyRuntimeState_ReInitThreads() if HAVE_FORK is defined. Said differently, put the whole function into #ifdef HAVE_FORK ... #endif. Here and where it's prototype in the header file (pycore_runtime.h).

Same remark for _PyImport_ReInitLock() and _PyEval_ReInitThreads().

**Review by vstinner — CHANGES_REQUESTED** [review]

**Comment by corona10** [comment]
I have made the requested changes; please review again

File: Include/internal/pycore_ceval.h [hunk 3]
```diff
@@ -25,8 +25,10 @@ PyAPI_FUNC(int) _PyEval_AddPendingCall(
     int (*func)(void *),
     void *arg);
 PyAPI_FUNC(void) _PyEval_SignalAsyncExc(PyThreadState *tstate);
+#ifdef HAVE_FORK
 PyAPI_FUNC(void) _PyEval_ReInitThreads(
     struct pyruntimestate *runtime);
```
Comment (vstinner) [inline Include/internal/pycore_ceval.h]: I don't think that this function must be exported. Please make it private with:

```suggestion
extern void _PyEval_ReInitThreads(struct pyruntimestate *runtime);
```

File: Python/pystate.c [hunk 4]
```diff
@@ -147,24 +149,24 @@ _PyRuntimeState_ReInitThreads(_PyRuntimeState *runtime)
...
     _PyMem_SetDefaultAllocator(PYMEM_DOMAIN_RAW, &old_alloc);
 
-    runtime->interpreters.mutex = PyThread_allocate_lock();
-    runtime->interpreters.main->id_mutex = PyThread_allocate_lock();
-    runtime->xidregistry.mutex = PyThread_allocate_lock();
-
-    PyMem_SetAllocator(PYMEM_DOMAIN_RAW, &old_alloc);
-
-    if (runtime->interpreters.mutex == NULL) {
+    if (_PyThread_at_fork_reinit(&runtime->interpreters.mutex) < 0) {
```
Comment (vstinner) [inline Python/pystate.c]: I suggest to add 3 local variables to store _PyThread_at_fork_reinit() result and only check the result after `PyMem_SetAllocator(PYMEM_DOMAIN_RAW, &old_alloc);` call. Just to limit code between the two PyMem calls.

File: Python/ceval.c [hunk 5]
```diff
@@ -397,6 +397,8 @@ PyEval_ReleaseThread(PyThreadState *tstate)
     drop_gil(&runtime->ceval, tstate);
 }
 
+#ifdef HAVE_FORK
+
```
Comment (vstinner) [inline Python/ceval.c]: nitpick: remove this empty line. Same remark for the 2 other functions :-p

**Review by vstinner — APPROVED** [review]
LGTM. You can merge it. Thanks, the updated code now looks better! It's good to no longer define the 3 functions on Windows where they make no sense.

**Pull request merged**